base/samples/Sonar.hpp

```
@@ -0,0 +1,436 @@
+        : time(time)
+        , bin_duration(bin_duration)
+        , beam_width(beam_width)
+        , beam_height(beam_height)
+        , speed_of_sound(getSpeedOfSoundInWater())
+        , bin_count(bin_count)
+        , beam_count(0)
+    {
+    }
+
+    Sonar(base::Time time, base::Time bin_duration, int bin_count, base::Angle beam_width, base::Angle beam_height,
+            int beam_count, bool per_beam_timestamps)
+        : time(time)
+        , bin_duration(bin_duration)
+        , beam_width(beam_width)
+        , beam_height(beam_height)
+        , speed_of_sound(getSpeedOfSoundInWater())
+        , bin_count(bin_count)
+        , beam_count(beam_count)
+    {
+        resize(bin_count, beam_count, per_beam_timestamps);
+    }
+
+    void resize(int bin_count, int beam_count, bool per_beam_timestamps)
+    {
+        if (per_beam_timestamps)
+            timestamps.resize(beam_count);
+
+        bins.resize(beam_count * bin_count, base::unknown<float>());
+        bearings.resize(beam_count, base::Angle::unknown());
+    }
+
+    /** Initializes a Sonar structure to represent a single beam
+     */
+    static Sonar fromSingleBeam(base::Time time, base::Time bin_duration, base::Angle beam_width, base::Angle beam_height,
+            std::vector<float> const& bins, base::Angle bearing = base::Angle())
+    {
+        Sonar sample(time, bin_duration, bins.size(), beam_width, beam_height);
+        sample.pushBeam(bins, bearing);
+        return sample;
+    }
+
+    /** The start of a bin in the time domain, relative to the beam's
+     * acquisition time
+     */
+    base::Time getBinRelativeStartTime(int bin_idx) const
+    {
+        return bin_duration * bin_idx;
+    }
+
+    /** The acquisition start of a beam
+     */
+    base::Time getBeamAcquisitionStartTime(int beam) const
+    {
+        if (timestamps.empty())
+            return timestamps.front();
```

This should either throw an exception or return the time field.